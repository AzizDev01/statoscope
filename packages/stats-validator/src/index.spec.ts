import os from 'os';
import Validator from './';

const output: string[][] = [];
let consoleLogSpy: jest.SpyInstance<void, string[]> | null = null;

beforeEach(() => {
  consoleLogSpy = jest
    .spyOn(global.console, 'log')
    .mockImplementation((...args: string[]) => {
      output.push(
        args.map((o) =>
          String(o).replace(process.cwd(), '<pwd>').replace(os.tmpdir(), '<tmp>')
        )
      );
    });
});

afterEach(() => {
  consoleLogSpy?.mockRestore();
  output.length = 0;
});

test('should work', async () => {
  const validator = new Validator({
    plugins: [[require.resolve('../../stats-validator-plugin-webpack'), 'webpack']],
    validate: {
      rules: {
        'webpack/restricted-modules': ['error', [/\.\/src/]],
        'webpack/restricted-packages': ['error', ['foo']],
      },
    },
  });

  const result = await validator.validate(
    require.resolve('../../../test/bundles/simple/stats-prod.json')
  );

  await validator.report(result);

  expect(result).toMatchSnapshot();
  expect(output).toMatchSnapshot();
});

test('custom reporter', async () => {
  const validator = new Validator({
    plugins: [[require.resolve('../../stats-validator-plugin-webpack'), 'webpack']],
    validate: {
      reporters: [
        [
          require.resolve('../../../test/fixtures/stats-validator/reporters/custom.js'),
          'foo',
        ],
      ],
      rules: {
        'webpack/restricted-modules': ['error', [/\.\/src/]],
        'webpack/restricted-packages': ['error', ['foo']],
      },
    },
  });

  const result = await validator.validate(
    require.resolve('../../../test/bundles/simple/stats-prod.json')
  );

  await validator.report(result);

  expect(result).toMatchSnapshot();
  expect(output).toMatchSnapshot();
});

test('silent', async () => {
  const validator = new Validator({
    plugins: [[require.resolve('../../stats-validator-plugin-webpack'), 'webpack']],
    validate: {
      reporters: [
        [
          require.resolve('../../../test/fixtures/stats-validator/reporters/custom.js'),
          'foo',
        ],
      ],
      silent: true,
      rules: {
        'webpack/restricted-modules': ['error', [/\.\/src/]],
        'webpack/restricted-packages': ['error', ['foo']],
      },
    },
  });

  const result = await validator.validate(
    require.resolve('../../../test/bundles/simple/stats-prod.json')
  );

  await validator.report(result);

  expect(result).toMatchSnapshot();
  expect(output).toMatchSnapshot();
});
