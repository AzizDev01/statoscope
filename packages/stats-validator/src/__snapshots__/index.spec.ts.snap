// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom reporter 1`] = `
Object {
  "files": Object {
    "input": "<pwd>/test/bundles/simple/stats-prod.json",
    "reference": undefined,
  },
  "rules": Array [
    Object {
      "api": Object {
        "getRuleDescriptor": [Function],
        "getStorage": [Function],
        "message": [Function],
        "setRuleDescriptor": [Function],
      },
      "execParams": Object {
        "mode": "error",
      },
      "name": "webpack/restricted-modules",
    },
    Object {
      "api": Object {
        "getRuleDescriptor": [Function],
        "getStorage": [Function],
        "message": [Function],
        "setRuleDescriptor": [Function],
      },
      "execParams": Object {
        "mode": "error",
      },
      "name": "webpack/restricted-packages",
    },
  ],
}
`;

exports[`should work 1`] = `
Object {
  "files": Object {
    "input": "<pwd>/test/bundles/simple/stats-prod.json",
    "reference": undefined,
  },
  "rules": Array [
    Object {
      "api": Object {
        "getRuleDescriptor": [Function],
        "getStorage": [Function],
        "message": [Function],
        "setRuleDescriptor": [Function],
      },
      "execParams": Object {
        "mode": "error",
      },
      "name": "webpack/restricted-modules",
    },
    Object {
      "api": Object {
        "getRuleDescriptor": [Function],
        "getStorage": [Function],
        "message": [Function],
        "setRuleDescriptor": [Function],
      },
      "execParams": Object {
        "mode": "error",
      },
      "name": "webpack/restricted-packages",
    },
  ],
}
`;

exports[`silent 1`] = `
Object {
  "files": Object {
    "input": "<pwd>/test/bundles/simple/stats-prod.json",
    "reference": undefined,
  },
  "rules": Array [
    Object {
      "api": Object {
        "getRuleDescriptor": [Function],
        "getStorage": [Function],
        "message": [Function],
        "setRuleDescriptor": [Function],
      },
      "execParams": Object {
        "mode": "error",
      },
      "name": "webpack/restricted-modules",
    },
    Object {
      "api": Object {
        "getRuleDescriptor": [Function],
        "getStorage": [Function],
        "message": [Function],
        "setRuleDescriptor": [Function],
      },
      "execParams": Object {
        "mode": "error",
      },
      "name": "webpack/restricted-packages",
    },
  ],
}
`;
